controllers/metal3remediation_controller.go

```diff
@@ -222,6 +250,203 @@ func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
+func (r *Metal3RemediationReconciler) remediateRebootStrategy(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface, node *corev1.Node) (ctrl.Result, error) {
+	switch remediationMgr.GetRemediationPhase() {
+	case infrav1.PhaseRunning:
+
+		// add finalizer
+		if set := remediationMgr.SetFinalizer(); set {
+			return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+		}
+
+		// power off if needed
+		if ok, err := remediationMgr.IsPowerOffRequested(ctx); err != nil {
+			r.Log.Error(err, "error getting poweroff annotation status")
+			return ctrl.Result{}, errors.Wrap(err, "error getting poweroff annotation status")
+		} else if !ok {
+			r.Log.Info("Powering off the host")
+			err = remediationMgr.SetPowerOffAnnotation(ctx)
+			if err != nil {
+				r.Log.Error(err, "error setting poweroff annotation")
+				return ctrl.Result{}, errors.Wrap(err, "error setting poweroff annotation")
+			}
+
+			// done for now, wait a bit before checking if we are powered off already
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		// wait until powered off
+		if on, err := remediationMgr.IsPoweredOn(ctx); err != nil {
+			r.Log.Error(err, "error getting power status")
+			return ctrl.Result{}, errors.Wrap(err, "error getting power status")
+		} else if on {
+			// wait a bit before checking again if we are powered off already
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		// if we have a node, store annotations and labels, and delete it
+		if node != nil {
+			/*
+				Delete the node only after the host is powered off. Otherwise, if we would delete the node
+				when the host is powered on, the scheduler would assign the workload to other nodes, with the
+				possibility that two instances of the same application are running in parallel. This might result
+				in corruption or other issues for applications with singleton requirement. After the host is powered
+				off we know for sure that it is safe to re-assign that workload to other nodes.
+			*/
+			modified := r.backupNode(ctx, remediationMgr, node)
+			if modified {
+				r.Log.Info("Backing up node")
+				// save annotations before deleting node
+				return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+			}
+			r.Log.Info("Deleting node")
+			err := remediationMgr.DeleteNode(ctx, node)
+			if err != nil {
+				r.Log.Error(err, "error deleting node")
+				return ctrl.Result{}, errors.Wrap(err, "error deleting node")
+			}
+			// wait until node is gone
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		// node is deleted: remove power annotation
+		ok, err := remediationMgr.IsPowerOffRequested(ctx)
+		if err != nil {
+			r.Log.Error(err, "error getting poweroff annotation status")
+			return ctrl.Result{}, errors.Wrap(err, "error getting poweroff annotation status")
+		} else if ok {
+			r.Log.Info("Powering on the host")
+			err := remediationMgr.RemovePowerOffAnnotation(ctx)
```

Comment:
You shouldn't do this until after the phase has been changed to Waiting, because if the write to the Status fails then we'll end up re-adding another reboot annotation.

Reply:
done